### DISCUSSION TITLE
Toggling Sounds + Controlling Cut Off

### DISCUSSION THREAD

**Discussion by garamoniker** [GeekFunkLabs/fluidpatcher#58] [General]
Hi guys,

I've had some amazing fun working with the Squishbox already (thanks, Bill!) and I'm ready to expand it's usage a little bit for my live band.

I've noticed in the premade script a suggestion of CCs 27+28 for toggling (I assume toggling through sounds). I've set up my controller (currently an Edirol PCR800) with these CCs on 2 buttons but it doesn't seem to be registering?

Originally, I wanted to simply map out different sounds button by button but, I've noticed, if I set up the controll for a Sound 1 button (from the scrollable list), then a Sound 2 button, e.t.c, it often seems to pull different sounds from the library instead?

Also, I'd love to hear a little bit about setting up a cut-off (wah) filter now, as well.

Thanks for your help, guys.

**Comment by albedozero** [comment]
The toggling in [bank1.yaml](https://github.com/albedozero/fluidpatcher/blob/435f432a8288ecced22132ee65733fc672065439/SquishBox/banks/bank1.yaml) is just for a couple patches. CC 27 toggles the octave doubling in the [Bass+Oct](https://github.com/albedozero/fluidpatcher/blob/435f432a8288ecced22132ee65733fc672065439/SquishBox/banks/bank1.yaml#L60) patch, and CC 28 toggles portamento/glide in the [Smooth Flute](https://github.com/albedozero/fluidpatcher/blob/435f432a8288ecced22132ee65733fc672065439/SquishBox/banks/bank1.yaml#L79) patch. The bank also looks for these CCs on MIDI channel 10, which is usually the case for drum pads.

Your Edirol is probably sending program change `prog` messages when set to Sound 1, Sound 2, etc. and this tells fluidsynth to select that program from whatever soundfont is loaded on that channel. If you want to switch between your patches, you can use a router rule with a `patch` parameter. There's a brief explanation of this in the [wiki](https://github.com/albedozero/fluidpatcher/blob/master/patcher/file_formats.md#router_rules), and I'll cover it in detail in a future lesson video. I've also got lesson videos coming up on effects plugins, which include adding an autowah, and controlling filter cutoff using NRPNs, etc. I know I'm slow getting videos out, so here's [my go-to bank file](https://gist.github.com/albedozero/817dd14370e6f49d90a2d5c27bdbb90d) I use for general playing situations, which has examples of all these. Feel free to ask questions about it.

**Reply by garamoniker** [reply]
Thanks, Bill. Sorry to annoy you with checking my work. I'm figuring that would look a bit like this:

GMBY
1: annoyingchiptunesound.sf2 000:000
router_rules:
{type: patch, chan: 1, par1: 19=1}
{type: patch, chan: 1, par1: 20=2}
{type: patch, chan: 1, par1: 21=3}

Bass Synth
1: terriblesh101sample.sf2 000:000
router_rules:
{type: patch, chan: 1, par1: 19=1}
{type: patch, chan: 1, par1: 20=2}
{type: patch, chan: 1, par1: 21=3}

Piano
1: tolerablepiano.sf2 000:000
router_rules:
{type: patch, chan: 1, par1: 19=1}
{type: patch, chan: 1, par1: 20=2}
{type: patch, chan: 1, par1: 21=3}

(Obviously, with indents included.) So, these 19, 20, 21 locations are 3 different buttons on my PCR. When I press the button on CC19, it should load patch 1(GMBY). When I press the button on CC20, it should load patch 2(Bass Synth). Would this be correct?

**Reply by albedozero** [reply]
Not quite - the rule type should still be `cc`, since that's what you're trying to catch. You want to add a `patch` _parameter_ with a value of `select`, like below, and have it route the value of 127 on channel 1 sent by pressing your momentary pad (I'm assuming) to a value corresponding to the patch index to select. Also, the patches start at zero, which I know is confusing b/c they start at 1 on the display - I need to fix that 😬. You could make it even cleaner by putting the router rules at bank level, so you wouldn't have to copy-paste them to every patch. I added patch up/down rules too, just to show how those work. No annoyance here - this is much quicker than making a video! 😃 

```yaml
patches:
  GMBY:
    1: annoyingchiptunesound.sf2 000:000
  Bass Synth:
    1: terriblesh101sample.sf2 000:000
  Piano:
    1: tolerablepiano.sf2 000:000
router_rules:
- {type: cc, chan: 1, par1: 19, patch: select, par2: 127=0} # selects first patch (GMBY)
- {type: cc, chan: 1, par1: 20, patch: select, par2: 127=1} # selects second patch (Bass Synth)
- {type: cc, chan: 1, par1: 21, patch: select, par2: 127=2} # selects third patch (Piano)
- {type: cc, chan: 1, par1: 22, patch: 1} # patch up
- {type: cc, chan: 1, par1: 23, patch: -1} # patch down
```

**Reply by garamoniker** [reply]
Thanks so much for your help!

I can see the cut off info on your go-to bank file. I'll have a play around and, if it's okay, I'm sure I'll come annoy you with that to when I get confused.

**Reply by garamoniker** [reply]
Ugh. So it's Wah time. I'm sorry. I opened up my Pi and input 

`sudo apt-get install tap-plugins wah-plugins`

In return, I get:

command not found. 

Basically, I'm just trying to install it as per your Synth setup - I've taken a look at the fluidsynth commands and I take it this is the only way to get control of cut-off/filter/resonance parameters.

**Reply by albedozero** [reply]
Actually, the better way to control the synthesis parameters like filter cutoff, resonance, amplitude/modulation envelopes, etc. is to use modulators. I explain what they are in [this section of Lesson 4](https://www.youtube.com/watch?v=K_TKv0kQubU&t=268s).

TLDR they let you change things with CC messages, but they don't exist in most soundfonts so you have to add them yourself. The ModSynth_R1.sf2 soundfont included with the SquishBox has a bunch of added modulators, so you can play around with [how they sound](https://youtube.com/shorts/QUQe0vcsR1w?feature=share), and download [Polyphone](https://www.polyphone-soundfonts.com/download) and open up the soundfont to see how I added them.

Kind of weird you got a `command not found` when trying to run apt-get. What happens if you just do a `sudo apt-get update`?

**Reply by garamoniker** [reply]
Ah! Awesome! Thanks, Bill. I've actually already been using Polyphone (hence the terrible soundfont names). I remember reading somewhere that fluid didn't support the Soundfont parameters. I'm very glad to be wrong.
